File: go-backend/models/transaction.go
package models

import (
	"database/sql"
	"time"
)

type Transaction struct {
	ID          int       `json:"id" db:"id"`
	UserID      int       `json:"user_id" db:"user_id"`
	GoalID      int       `json:"goal_id" db:"goal_id"`
	Amount      float64   `json:"amount" db:"amount"`
	Description string    `json:"description" db:"description"`
	Type        string    `json:"type" db:"type"` // "add" or "remove"
	CreatedAt   time.Time `json:"created_at" db:"created_at"`
}

type TransactionRepository struct {
	db *sql.DB
}

func NewTransactionRepository(db *sql.DB) *TransactionRepository {
	return &TransactionRepository{db: db}
}

func (r *TransactionRepository) Create(transaction *Transaction) error {
	query := `
		INSERT INTO transactions (user_id, goal_id, amount, description, type, created_at)
		VALUES (?, ?, ?, ?, ?, ?)
	`
	result, err := r.db.Exec(query, transaction.UserID, transaction.GoalID, transaction.Amount, transaction.Description, transaction.Type, time.Now())
	if err != nil {
		return err
	}
	
	id, err := result.LastInsertId()
	if err != nil {
		return err
	}
	
	transaction.ID = int(id)
	return nil
}

func (r *TransactionRepository) GetByGoalID(goalID int) ([]Transaction, error) {
	query := `
		SELECT id, user_id, goal_id, amount, description, type, created_at
		FROM transactions
		WHERE goal_id = ?
		ORDER BY created_at DESC
	`
	rows, err := r.db.Query(query, goalID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var transactions []Transaction
	for rows.Next() {
		var transaction Transaction
		err := rows.Scan(
			&transaction.ID, &transaction.UserID, &transaction.GoalID,
			&transaction.Amount, &transaction.Description, &transaction.Type,
			&transaction.CreatedAt,
		)
		if err != nil {
			return nil, err
		}
		transactions = append(transactions, transaction)
	}
	return transactions, nil
}

func (r *TransactionRepository) GetByUserID(userID int) ([]Transaction, error) {
	query := `
		SELECT id, user_id, goal_id, amount, description, type, created_at
		FROM transactions
		WHERE user_id = ?
		ORDER BY created_at DESC
	`
	rows, err := r.db.Query(query, userID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var transactions []Transaction
	for rows.Next() {
		var transaction Transaction
		err := rows.Scan(
			&transaction.ID, &transaction.UserID, &transaction.GoalID,
			&transaction.Amount, &transaction.Description, &transaction.Type,
			&transaction.CreatedAt,
		)
		if err != nil {
			return nil, err
		}
		transactions = append(transactions, transaction)
	}
	return transactions, nil
}

func (r *TransactionRepository) GetTotalByGoalID(goalID int) (float64, error) {
	query := `
		SELECT 
			COALESCE(SUM(CASE WHEN type = 'add' THEN amount ELSE -amount END), 0) as total
		FROM transactions
		WHERE goal_id = ?
	`
	var total float64
	err := r.db.QueryRow(query, goalID).Scan(&total)
	return total, err
}

File: flutter_app/lib/models/transaction.dart
class Transaction {
  final int id;
  final int userId;
  final int goalId;
  final double amount;
  final String description;
  final String type; // "add" or "remove"
  final DateTime createdAt;

  Transaction({
    required this.id,
    required this.userId,
    required this.goalId,
    required this.amount,
    required this.description,
    required this.type,
    required this.createdAt,
  });

  factory Transaction.fromJson(Map<String, dynamic> json) {
    return Transaction(
      id: json['id'] as int,
      userId: json['user_id'] as int,
      goalId: json['goal_id'] as int,
      amount: (json['amount'] as num).toDouble(),
      description: json['description'] as String,
      type: json['type'] as String,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'goal_id': goalId,
      'amount': amount,
      'description': description,
      'type': type,
      'created_at': createdAt.toIso8601String(),
    };
  }

  bool get isAddition => type == 'add';
  bool get isRemoval => type == 'remove';
}
